File: clients/windows/ServerLoad.cs
using System.Net.Http;
using System.Text.Json;
namespace FamilyConnect;
internal sealed record LoadSample(string Country,double Observed,double Cpu,double Rx,double Tx,double? Percent,bool Estimated)
{
    internal bool Fresh=>FreshAt(DateTimeOffset.UtcNow.ToUnixTimeSeconds());
    internal bool FreshAt(double now)=>now-Observed is >=-15 and <=45;
}
internal static class ServerLoad
{
    static readonly HttpClient http=new(new HttpClientHandler{AllowAutoRedirect=false}){Timeout=TimeSpan.FromSeconds(5),MaxResponseContentBufferSize=8192};
    internal static string? Country(string? endpoint)=>endpoint switch{"185.251.89.19"=>"ru","186.246.45.246"=>"nl",_=>null};
    internal static LoadSample Parse(string raw,string country,double now)
    {
        if(raw.Length>8192||country is not ("ru" or "nl"))throw new FormatException();
        using var doc=JsonDocument.Parse(raw);var root=doc.RootElement;
        if(root.GetProperty("schema").GetInt32()!=1)throw new FormatException();
        var o=root.GetProperty("gateways").GetProperty(country);
        if(o.GetProperty("country").GetString()!=country)throw new FormatException();
        double Number(string name,double min=0,double max=1e12){var e=o.GetProperty(name);if(e.ValueKind!=JsonValueKind.Number)throw new FormatException();double n=e.GetDouble();if(!double.IsFinite(n)||n<min||n>max)throw new FormatException();return n;}
        double at=Number("observed_at"),cpu=Number("cpu_percent",0,100),rx=Number("rx_mbps"),tx=Number("tx_mbps");
        string? direction=o.TryGetProperty("capacity_direction",out var d)?d.GetString():"duplex";
        if(direction is not ("duplex" or "egress"))throw new FormatException();
        double? percent=null;if(o.TryGetProperty("capacity_mbps",out var cap)&&cap.ValueKind!=JsonValueKind.Null)
            percent=Math.Min(100,Math.Max(cpu,100*(direction=="egress"?tx:Math.Max(rx,tx))/Number("capacity_mbps",.001,1e9)));
        bool estimated=o.TryGetProperty("capacity_basis",out var b)&&b.GetString()=="provider-default-estimate";
        var result=new LoadSample(country,at,cpu,rx,tx,percent,estimated);if(!result.FreshAt(now))throw new FormatException();return result;
    }
    internal static async Task<LoadSample> Fetch(string country)=>Parse(await http.GetStringAsync("https://185.251.89.19:8443/status/server-load.json"),country,DateTimeOffset.UtcNow.ToUnixTimeSeconds());
    internal static async Task<Dictionary<string,LoadSample>> FetchAll()
    {
        var raw=await http.GetStringAsync("https://185.251.89.19:8443/status/server-load.json");
        var now=DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        using var doc=JsonDocument.Parse(raw);var root=doc.RootElement;
        if(root.GetProperty("schema").GetInt32()!=1)throw new FormatException();
        var result=new Dictionary<string,LoadSample>();
        foreach(var country in new[]{"ru","nl"})
            if(root.GetProperty("gateways").TryGetProperty(country,out _))
                try{result[country]=Parse(raw,country,now);}catch(FormatException){}
        if(result.Count==0)throw new FormatException();
        return result;
    }
    internal static void Check()
    {
        const string json="""{"schema":1,"gateways":{"nl":{"country":"nl","observed_at":1000,"cpu_percent":30,"rx_mbps":900,"tx_mbps":100,"capacity_mbps":200,"capacity_direction":"egress","capacity_basis":"provider-default-estimate"}}}""";
        var s=Parse(json,"nl",1010);if(s.Percent!=50||!s.Estimated)throw new Exception("Load calculation");
        if(Parse(json.Replace("\"capacity_mbps\":200","\"capacity_mbps\":null"),"nl",1010).Percent is not null)throw new Exception("Unknown capacity");
        foreach(var now in new[]{1046d,980d}){bool rejected=false;try{Parse(json,"nl",now);}catch(FormatException){rejected=true;}if(!rejected)throw new Exception("Stale load accepted");}
        if(Country("127.0.0.1") is not null)throw new Exception("Unknown server mapped");
    }
}
internal sealed class LoadBar:Control
{
    internal double? Percent;
    public LoadBar(){DoubleBuffered=true;ResizeRedraw=true;Height=8;Dock=DockStyle.Fill;Margin=new Padding(0,4,0,8);AccessibleRole=AccessibleRole.ProgressBar;}
    protected override void OnPaint(PaintEventArgs e){base.OnPaint(e);using var border=new Pen(Color.FromArgb(67,142,121));e.Graphics.DrawRectangle(border,0,0,Math.Max(0,Width-1),Math.Max(0,Height-1));if(Percent is double p){using var fill=new SolidBrush(p>=90?Color.FromArgb(239,132,116):p>=70?Color.FromArgb(255,173,70):Color.FromArgb(152,247,216));e.Graphics.FillRectangle(fill,1,1,(int)((Width-2)*Math.Clamp(p,0,100)/100),Math.Max(0,Height-2));}}
}

File: clients/android/app/src/main/java/com/familyconnect/app/ServerLoad.java
package com.familyconnect.app;

import com.google.gson.*;
import java.net.URL;
import javax.net.ssl.HttpsURLConnection;
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.*;

/** Informational host utilization, never a VPN health or routing decision. */
final class ServerLoad {
    final String country;final double observed,cpu,rx,tx;final Double percent;final boolean estimated;
    ServerLoad(String c,double t,double p,double r,double x,Double v,boolean e){country=c;observed=t;cpu=p;rx=r;tx=x;percent=v;estimated=e;}
    boolean fresh(double now){return now-observed>=-15&&now-observed<=45;}
    static double number(JsonObject o,String name,double min,double max){
        JsonElement e=o.get(name);if(e==null||!e.isJsonPrimitive()||!e.getAsJsonPrimitive().isNumber())throw new IllegalArgumentException();
        double v=e.getAsDouble();if(!Double.isFinite(v)||v<min||v>max)throw new IllegalArgumentException();return v;
    }
    private static ServerLoad gateway(JsonObject o,String country,double now){
        if(!country.equals(o.get("country").getAsString()))throw new IllegalArgumentException();
        double t=number(o,"observed_at",0,1e12),cpu=number(o,"cpu_percent",0,100),rx=number(o,"rx_mbps",0,1e12),tx=number(o,"tx_mbps",0,1e12);
        String direction=o.has("capacity_direction")?o.get("capacity_direction").getAsString():"duplex";
        if(!direction.equals("duplex")&&!direction.equals("egress"))throw new IllegalArgumentException();
        Double value=null;
        if(o.has("capacity_mbps")&&!o.get("capacity_mbps").isJsonNull())value=Math.min(100,Math.max(cpu,100*(direction.equals("egress")?tx:Math.max(rx,tx))/number(o,"capacity_mbps",.001,1e9)));
        boolean estimated=o.has("capacity_basis")&&"provider-default-estimate".equals(o.get("capacity_basis").getAsString());
        ServerLoad result=new ServerLoad(country,t,cpu,rx,tx,value,estimated);if(!result.fresh(now))throw new IllegalArgumentException();return result;
    }
    static ServerLoad parse(String raw,String country,double now){
        if(raw.length()>8192||!(country.equals("ru")||country.equals("nl")))throw new IllegalArgumentException();
        JsonObject root=JsonParser.parseString(raw).getAsJsonObject();if(number(root,"schema",1,1)!=1)throw new IllegalArgumentException();
        JsonObject gateways=root.getAsJsonObject("gateways");if(!gateways.has(country))throw new IllegalArgumentException();
        return gateway(gateways.getAsJsonObject(country),country,now);
    }
    static Map<String,ServerLoad> fetchAll()throws IOException {
        HttpsURLConnection c=(HttpsURLConnection)new URL("https://185.251.89.19:8443/status/server-load.json").openConnection();
        c.setConnectTimeout(5000);c.setReadTimeout(5000);c.setInstanceFollowRedirects(false);c.setUseCaches(false);
        try{if(c.getResponseCode()!=200)throw new IOException();
            ByteArrayOutputStream out=new ByteArrayOutputStream();byte[] buffer=new byte[1024];
            try(InputStream in=c.getInputStream()){int n;while((n=in.read(buffer))!=-1){out.write(buffer,0,n);if(out.size()>8192)throw new IOException();}}
            JsonObject root=JsonParser.parseString(out.toString(StandardCharsets.UTF_8.name())).getAsJsonObject();
            if(number(root,"schema",1,1)!=1)throw new IllegalArgumentException();
            JsonObject gateways=root.getAsJsonObject("gateways");double now=System.currentTimeMillis()/1000.0;
            Map<String,ServerLoad> result=new LinkedHashMap<>();
            for(String country:new String[]{"ru","nl"}){
                if(!gateways.has(country))continue;
                try{result.put(country,gateway(gateways.getAsJsonObject(country),country,now));}catch(IllegalArgumentException ignored){}
            }
            if(result.isEmpty())throw new IOException();return result;
        }finally{c.disconnect();}
    }
    static ServerLoad fetch(String country)throws IOException {
        HttpsURLConnection c=(HttpsURLConnection)new URL("https://185.251.89.19:8443/status/server-load.json").openConnection();
        c.setConnectTimeout(5000);c.setReadTimeout(5000);c.setInstanceFollowRedirects(false);c.setUseCaches(false);
        try{if(c.getResponseCode()!=200)throw new IOException();
            ByteArrayOutputStream out=new ByteArrayOutputStream();byte[] buffer=new byte[1024];
            try(InputStream in=c.getInputStream()){int n;while((n=in.read(buffer))!=-1){out.write(buffer,0,n);if(out.size()>8192)throw new IOException();}}
            return parse(out.toString(StandardCharsets.UTF_8.name()),country,System.currentTimeMillis()/1000.0);
        }finally{c.disconnect();}
    }
}
